### PR TITLE
#417

### PR DESCRIPTION
**Pull request by HDouss**
Resolving issue #417 
Added LCOM4 test cases.
(fixed expected value for MethodMethodCalls)

### CODE REVIEW

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) everybody who has role `REV` is banned at [#423](https://github.com/yegor256/jpeek/issues/423); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

**Comment by paulodamaso** [comment]
@0crat assign @fanifieiev

**Comment by 0crat** [comment]
> @0crat assign @fanifieiev  ([here](https://github.com/yegor256/jpeek/issues/423#issuecomment-597648951))

@paulodamaso This pull request [#423](https://github.com/yegor256/jpeek/issues/423) is assigned to @fanifieiev[/z](https://www.0crat.com/u/fanifieiev), here is [why](https://www.0crat.com/footprint/C7JGJ00DP/e21550dc-f936-4645-9e0a-b72fe6578aba); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/org/jpeek/calculus/java/Lcom4Test.java [hunk 1]
```diff
@@ -65,18 +64,19 @@ public void createsXmlCalculusWithXpaths() throws IOException {
         ).affirm();
     }
 
-    @Test
+    @ParameterizedTest
     @Disabled
-    public void methodMethodCalls() throws Exception {
+    @CsvFileSource(resources = "/org/jpeek/calculus/java/lcom4-params.csv")
+    public void methodMethodCalls(final String file, final String value) throws Exception {
```
Comment (fanifieiev) [inline src/test/java/org/jpeek/calculus/java/Lcom4Test.java]: @HDouss I know that is not your change, but please change the name of the method to a meaningful one

File: src/test/java/org/jpeek/calculus/java/Lcom4Test.java [hunk 2]
```diff
@@ -65,18 +64,19 @@ public void createsXmlCalculusWithXpaths() throws IOException {
...
         final XML result = new Lcom4().node(
```
Comment (fanifieiev) [inline src/test/java/org/jpeek/calculus/java/Lcom4Test.java]: @HDouss I think `Lcom4` should have a constructor or several constructors to accept parameters and  the only its interfaced method `node` should be parameterless.
I mean, `Calculus ` should have:
```
public interface Calculus {
   
    XML node()
        throws IOException;

}
```

and `Lcom4` should have a cosntructor to accept the values.

File: src/test/java/org/jpeek/calculus/java/Lcom4Test.java [hunk 3]
```diff
@@ -65,18 +64,19 @@ public void createsXmlCalculusWithXpaths() throws IOException {
...
             "", new HashMap<>(0), new Skeleton(
-                new FakeBase("MethodMethodCalls")
+                new FakeBase(file)
             ).xml()
         );
         new Assertion<>(
             "Must create LCOM4 value",
             result.xpath("/metric/app/package/class/@value").get(0),
```
Comment (fanifieiev) [inline src/test/java/org/jpeek/calculus/java/Lcom4Test.java]: @HDouss The Cactoos has `ItemAt` scalar, so you could use it like an elegant way:
`new ItemAt(0, result.xpath("/metric/app/package/class/@value"))`
and test scalars

**Comment by HDouss** [comment]
@fanifieiev Actually all your 3 comments are not related to the issue or the PR. I fixed comment 1&3. The comment 2 (about node interface method): I am not sure this is a good idea + this is a rather big refactoring as it will affect xsl calculus as well, along their usages: metrics classes, App class, Test classes... Please submit a new issue so the ARC could state if it is a good idea (also because it is unrelated to that issue).

**Comment by fanifieiev** [comment]
> @fanifieiev Actually all your 3 comments are not related to the issue or the PR. I fixed comment 1&3. The comment 2 (about node interface method): I am not sure this is a good idea + this is a rather big refactoring as it will affect xsl calculus as well, along their usages: metrics classes, App class, Test classes... Please submit a new issue so the ARC could state if it is a good idea (also because it is unrelated to that issue).

@HDouss I know it might not be related to this PR, but, reviewing even the old code, help the reviewer detect potential design problems and ask DEV to leave a todo.
So, I personally think, that making such comments is a good practice. 
@paulodamaso WDYT???

**Comment by HDouss** [comment]
@paulodamaso ping

**Comment by paulodamaso** [comment]
> > @fanifieiev Actually all your 3 comments are not related to the issue or the PR. I fixed comment 1&3. The comment 2 (about node interface method): I am not sure this is a good idea + this is a rather big refactoring as it will affect xsl calculus as well, along their usages: metrics classes, App class, Test classes... Please submit a new issue so the ARC could state if it is a good idea (also because it is unrelated to that issue).
> 
> @HDouss I know it might not be related to this PR, but, reviewing even the old code seen in PRs, helps the reviewer detect potential design problems and ask DEV to leave a todo.
> So, I personally think, that making such comments is a good practice.
> @paulodamaso WDYT???

@fanifieiev In this case I think that it's better that you open an issue and get the payment for bug report. Asking the DEV to leave a puzzle don't earn you the payment for this bug.

**Comment by fanifieiev** [comment]
> > > @fanifieiev Actually all your 3 comments are not related to the issue or the PR. I fixed comment 1&3. The comment 2 (about node interface method): I am not sure this is a good idea + this is a rather big refactoring as it will affect xsl calculus as well, along their usages: metrics classes, App class, Test classes... Please submit a new issue so the ARC could state if it is a good idea (also because it is unrelated to that issue).
> > 
> > 
> > @HDouss I know it might not be related to this PR, but, reviewing even the old code seen in PRs, helps the reviewer detect potential design problems and ask DEV to leave a todo.
> > So, I personally think, that making such comments is a good practice.
> > @paulodamaso WDYT???
> 
> @fanifieiev In this case I think that it's better that you open an issue and get the payment for bug report. Asking the DEV to leave a puzzle don't earn you the payment for this bug.
> 
> Usually we ask the performer to leave puzzles when the budget for the PR was already spent (30 min) but there are some features still missing.

@paulodamaso OK)))

File: src/test/java/org/jpeek/calculus/java/Lcom4Test.java [hunk 4]
```diff
@@ -65,18 +65,19 @@ public void createsXmlCalculusWithXpaths() throws IOException {
...
-                new FakeBase("MethodMethodCalls")
+                new FakeBase(file)
             ).xml()
         );
         new Assertion<>(
             "Must create LCOM4 value",
-            result.xpath("/metric/app/package/class/@value").get(0),
-            new IsEqual<>("1")
+            new ItemAt<>(0, result.xpath("/metric/app/package/class/@value")).value(),
+            new IsEqual<>(value)
```
Comment (fanifieiev) [inline src/test/java/org/jpeek/calculus/java/Lcom4Test.java]: @HDouss Please use `ScalarHasValue`

Comment (HDouss) [reply src/test/java/org/jpeek/calculus/java/Lcom4Test.java]: @fanifieiev fixed

**Comment by HDouss** [comment]
@fanifieiev Please review again

**Comment by HDouss** [comment]
@paulodamaso ping

**Comment by paulodamaso** [comment]
@HDouss Thank you

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by HDouss** [comment]
@paulodamaso seems we still have problems with rultor

**Comment by paulodamaso** [comment]
@HDouss Yes, I'm trying to fix it since last week.

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by paulodamaso** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job was finished in 27 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Pull request merged**